shark_turbine/kernel/wave/thread_shape_analysis.py

```diff
@@ -0,0 +1,118 @@
+        if target.index not in custom.index:
+            raise NotImplementedError(
+                "NYI: Handle when source target index size is not found in target/user index."
+            )
+        custom.index[target.index].size = target.size
+
+
+# Function called on op post propagation for extra processing/handling.
+def post_propagation(custom: CustomOp, target_index_sizes: list[IndexSize]):
+    if isinstance(custom, IterArg):
+        init_args = custom.parent_op().init_args[custom.get_iter_idx()]
+        set_index_size(get_custom(init_args), target_index_sizes)
+
+
+def determine_thread_shapes(trace: CapturedTrace):
+    """
+    This function does analysis and propagation of thread shape. It does by such:
+    1. Look for "anchor" ops who has information of it's elem_per_thread.
+    2. Do a forward/backward slice on these anchor ops to get ops that
+    who's shapes depends on these anchor ops.
+    3. We bucket these ops to Variadic(Index->elem_per_thread) mapping.
+    4. At every bucket of (index -> elem_per_thread), we apply these information
+       by updating their indexSequence size.
+    """
+
+    # Anchor ops are ops who's thread shape are predetermined.
+    anchorOpTypes = (Read, Write, MMA, ReduceOp)
+    noHandleTypes = (Placeholder, Output, ExtractSlice, Allocate)
+    nonPropagatableTypes = anchorOpTypes + noHandleTypes
+
+    def is_anchor_op(node: fx.Node):
+        return isinstance(get_custom(node), anchorOpTypes)
+
+    def propagatable_op(node: fx.Node):
+        return not isinstance(get_custom(node), nonPropagatableTypes)
+
+    anchor_ops = trace.walk(is_anchor_op)
+    thread_size_to_ops: dict[IndexSymbol, set[CustomOp]] = {}
+    for anchor_op in anchor_ops:
+        custom = get_custom(anchor_op)
+        index_sizes = get_custom_index_sizes(custom)
+        if isinstance(custom, (Read, ReduceOp)):
+            fwd_slice = capture_forward_slice(custom.fx_node, propagatable_op)
+            fwd_slice.remove(custom.fx_node)
+            thread_size_to_ops[index_sizes] = thread_size_to_ops.get(
+                index_sizes, set([])
+            ).union(fwd_slice)
+        elif isinstance(custom, Write):
+            bwd_slice = capture_backward_slice(custom.fx_node, propagatable_op)
+            bwd_slice.remove(custom.fx_node)
+            thread_size_to_ops[index_sizes] = thread_size_to_ops.get(
+                index_sizes, set([])
+            ).union(bwd_slice)
+        elif isinstance(custom, MMA):
+            lhs_bwd_slice = capture_backward_slice(custom.lhs, propagatable_op)
+            rhs_bwd_slice = capture_backward_slice(custom.rhs, propagatable_op)
+            acc_slice = capture_forward_slice(custom.acc, propagatable_op)
+            acc_slice.union(capture_backward_slice(custom.acc, propagatable_op))
+            acc_index = get_index_sizes(custom.acc_index)
+            lhs_index = get_index_sizes(custom.lhs_index)
+            rhs_index = get_index_sizes(custom.rhs_index)
```

Comment:
Just curious, how come we have to remove custom.fx_node from the backward slice of Write and forward slice of Read, but for MMA we don't need to remove anything from the forward or backward slices? Basically why dont we have anything like
```
bwd_slice.remove(custom.fx_node)
```
in the MMA case?

Reply:
haha we don't have to, I guess because it already have the index size so just slightly redundant to set the index size again. But I think you're right, it's not very expensive to set index and would be cleaner without it. :)

Reply:
done!